File: Sources/windows/BitTherapyCLR/src/capabilities/bounce_when_lateral_bound_is_hit.cpp
#include "bounce_when_lateral_bound_is_hit.h"

#include <chrono>
#include <iostream>

#include "../game/game.h"

BounceWhenLateralBoundIsHit::BounceWhenLateralBoundIsHit(double leftBound, double rightBound) : 
    leftBound(leftBound),
    rightBound(rightBound)
{}

void BounceWhenLateralBoundIsHit::update(std::chrono::milliseconds timeSinceLastUpdate, Entity * entity) {
    if (entity->frame.x < leftBound && entity->direction.x < 0) {
        entity->direction = Vector2d(-1 * entity->direction.x, entity->direction.y);
        entity->frame.x = leftBound;
    } 
    if (entity->frame.maxX() > rightBound && entity->direction.x > 0) {
        entity->direction = Vector2d(-1 * entity->direction.x, entity->direction.y);
        entity->frame.x = rightBound - entity->frame.w;
    }
}

File: Sources/windows/App/Sources/Capabilities/BounceWhenLateralBoundIsHit.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace App
{
    public class BounceWhenLateralBoundIsHit : EntityCapability
    {
        private double leftBound;
        private double rightBound;

        public BounceWhenLateralBoundIsHit(double leftBound, double rightBound)
        {
            this.leftBound = leftBound;
            this.rightBound = rightBound;
        }

        public override void Update(TimeSpan timeSinceLastUpdate, Entity entity)
        {
            if (entity.Frame.X < leftBound && entity.Direction.X < 0)
            {
                entity.Direction = new Vector2d(-1 * entity.Direction.X, entity.Direction.Y);
                entity.Frame.X = leftBound;
            }

            if (entity.Frame.MaxX() > rightBound && entity.Direction.X > 0)
            {
                entity.Direction = new Vector2d(-1 * entity.Direction.X, entity.Direction.Y);
                entity.Frame.X = rightBound - entity.Frame.W;
            }
        }
    }
}
